src/vs/gitpod/node/gitpodInsightsAppender.ts

```diff
@@ -0,0 +1,229 @@
+	readonly token: TokenServiceClient;
+	readonly info: InfoServiceClient;
+
+	constructor() {
+		this.status = new StatusServiceClient(this.addr, grpc.credentials.createInsecure());
+		this.token = new TokenServiceClient(this.addr, grpc.credentials.createInsecure());
+		this.info = new InfoServiceClient(this.addr, grpc.credentials.createInsecure());
+	}
+}
+
+type GitpodConnection = Omit<GitpodServiceImpl<GitpodClient, GitpodServer>, 'server'> & {
+	server: Pick<GitpodServer, 'trackEvent'>
+};
+
+async function getSupervisorData() {
+	const supervisor = new SupervisorConnection();
+
+	let contentAvailable = false;
+	while (!contentAvailable) {
+		try {
+			const contentStatusRequest = new ContentStatusRequest();
+			contentStatusRequest.setWait(true);
+			const result = await util.promisify(supervisor.status.contentStatus.bind(supervisor.status, contentStatusRequest, supervisor.metadata, {
+				deadline: Date.now() + supervisor.deadlines.long
+			}))();
+			contentAvailable = result.getAvailable();
+		} catch (e) {
+			console.error('Cannot maintain connection to supervisor', e);
+		}
+	}
+
+	const workspaceInfo = await util.promisify(supervisor.info.workspaceInfo.bind(supervisor.info, new WorkspaceInfoRequest(), supervisor.metadata, {
+		deadline: Date.now() + supervisor.deadlines.long
+	}))();
+	const gitpodApi = workspaceInfo.getGitpodApi()!;
+	const gitpodApiHost = gitpodApi.getHost();
+	const gitpodApiEndpoint = gitpodApi.getEndpoint();
+	const gitpodHost = workspaceInfo.getGitpodHost();
+	const workspaceId = workspaceInfo.getWorkspaceId();
+	const instanceId = workspaceInfo.getInstanceId();
+
+	const getTokenRequest = new GetTokenRequest();
+	getTokenRequest.setKind('gitpod');
+	getTokenRequest.setHost(gitpodApiHost);
+	getTokenRequest.addScope('function:trackEvent');
+
+	const getTokenResponse = await util.promisify(supervisor.token.getToken.bind(supervisor.token, getTokenRequest, supervisor.metadata, {
+		deadline: Date.now() + supervisor.deadlines.long
+	}))();
+	const serverToken = getTokenResponse.getToken();
+
+	return {
+		serverToken,
+		gitpodHost,
+		gitpodApiEndpoint,
+		workspaceId,
+		instanceId
+	};
+}
+
+async function getClient(serverToken: string, gitpodHost: string, gitpodApiEndpoint: string): Promise<GitpodConnection> {
+	const factory = new JsonRpcProxyFactory<GitpodServer>();
+	const gitpodService = new GitpodServiceImpl<GitpodClient, GitpodServer>(factory.createProxy()) as GitpodConnection;
+
+	const webSocket = new (ReconnectingWebSocket as any)(gitpodApiEndpoint, undefined, {
+		maxReconnectionDelay: 10000,
+		minReconnectionDelay: 1000,
+		reconnectionDelayGrowFactor: 1.3,
+		connectionTimeout: 10000,
+		maxRetries: Infinity,
+		debug: false,
+		startClosed: false,
+		WebSocket: class extends WebSocket {
+			constructor(address: string, protocols?: string | string[]) {
+				super(address, protocols, {
+					headers: {
+						'Origin': new URL(gitpodHost).origin,
+						'Authorization': `Bearer ${serverToken}`
```

Comment:
@jeanp413 we need to add here `User-Agent` at least. It should be a requirement for all our clients.